doc/rbd: refine rbd-snapshot.rst (#49394)
ceph/ceph: doc/rbd/rbd-snapshot.rst

.. index:: Ceph Block Device; snapshots

A snapshot is a read-only logical copy of an image at a particular point in
time: a checkpoint. One of the advanced features of Ceph block devices is that
you can create snapshots of images to retain point-in-time state history.  Ceph
also supports snapshot layering, which allows you to clone images (for example,
VM images) quickly and easily. Ceph block device snapshots are managed using
the ``rbd`` command and several higher-level interfaces, including `QEMU`_,
`libvirt`_, `OpenStack`_, and `CloudStack`_.

.. important:: To use RBD snapshots, you must have a running Ceph cluster.

> .. note:: Because RBD is unaware of any file system within an image (volume),

Reviewer: RBD is unaware of the contents of the image and the "thing" that is using the image: it could be a filesystem (potentially a root filesystem if RBD is backing a VM), someone using a block device directly, etc.  The notion of consistency is specific to that "thing".  A basic crash consistency, filesystem-level consistency and application-level consistency are distinguished (at least).

> Does this mean to say that snapshots are crash-consistent only if they are not co-ordinated within the mounting operating system?

No, RBD snapshots should always be crash-consistent.

> Also, what does it mean to be co-ordinated within that operating system?

It means letting the operating system know that the snapshot is about to be taken so that it can take any actions necessary to ensure required consistency (notify applications, quiesce filesystems, etc).

> .. note:: Because RBD is unaware of any file system within an image (volume),

Reviewer: I think this statement is correct as is:
- by default (no coordination), snapshots are crash-consistent
- depending on the level of coordination, snapshot can become filesystem-level or application-level consistent

Crash consistency is the weakest level of consistency, hence the use of "only".

> .. note:: Because RBD is unaware of any file system within an image (volume),

Reviewer: Sure!

> .. note:: Because RBD is unaware of any file system within an image (volume),

Reviewer 2: OK, thanks!

When `cephx`_ authentication is enabled (it is by default), you must specify a
user name or ID and a path to the keyring containing the corresponding key. See
:ref:`User Management <user-management>` for details. 

.. prompt:: bash $

	rbd --id {user-ID} --keyring /path/to/secret [commands]
	rbd --name {username} --keyring /path/to/secret [commands]

For example:

.. prompt:: bash $

	rbd --id admin --keyring=/etc/ceph/ceph.keyring [commands]

> 	rbd --name client.admin --keyring=/etc/ceph/ceph.keyring [commands]

Reviewer: Need to get rid of `=` here too.

Author: I thought I got these. Anyway, I got them now.

When `cephx`_ authentication is enabled (it is by default), you must specify a
user name or ID and a path to the keyring containing the corresponding key. See
:ref:`User Management <user-management>` for details. You may also set the
``CEPH_ARGS`` environment variable to avoid re-entry of these parameters:

.. prompt:: bash $

	rbd --id {user-ID} --keyring=/path/to/secret [commands]

> 	rbd --name {username} --keyring=/path/to/secret [commands]

Reviewer: Let's make this `--keyring /path/to/secret` (without `=`), so that it looks like all other options.

For example:

.. prompt:: bash $

	rbd --id admin --keyring=/etc/ceph/ceph.keyring [commands]
	rbd --name client.admin --keyring=/etc/ceph/ceph.keyring [commands]

.. tip:: Add the user and secret to the ``CEPH_ARGS`` environment variable to

>    avoid re-entry of these parameters.

Reviewer: This tip pretty much duplicates the last sentence of the previous paragraph.  I'd drop it and keep the tip as the tip gets nicely highlighted.

Author: Done. I removed the sentence and kept the tip.

Create Snapshot
---------------

> To create a snapshot use the ``rbd snap create`` command and specify the pool

Reviewer: ```suggestion
To create a snapshot, use the ``rbd snap create`` command and specify the pool
```

> name and the image name:

Reviewer: `rbd snap create` won't work without the desired snap name: "the pool name, the image name, and the snap name".

Author: Done.

.. prompt:: bash $

   rbd snap create {pool-name}/{image-name}@{snap-name}

For example:

.. prompt:: bash $

   rbd snap create rbd/foo@snapname
	

List Snapshots
--------------

To list the snapshots of an image, use the ``rbd snap ls`` command and specify
the pool name and the image name:

.. prompt:: bash $

   rbd snap ls {pool-name}/{image-name}

For example:

.. prompt:: bash $

   rbd snap ls rbd/foo

Roll back Snapshot
------------------

To roll back to a snapshot,  use the ``rbd snap rollback`` command and specify
the pool name, the image name, and the snap name:

.. prompt:: bash $

   rbd snap rollback {pool-name}/{image-name}@{snap-name}

For example:

.. prompt:: bash $

   rbd snap rollback rbd/foo@snapname

.. note:: Rolling back an image to a snapshot means overwriting the current
   version of the image with data from a snapshot. The time it takes to execute
   a rollback increases with the size of the image. It is **faster to clone**

>    from a snapshot **than to rollback** an image to a snapshot. Cloning from a

Reviewer: Since you started differentiating between "roll back" and "rollback", I think it's a verb here?
```suggestion
   from a snapshot **than to roll back** an image to a snapshot. Cloning from a
```

Author: It absolutely is. This change is in.

   snapshot is the preferred method of returning to a pre-existing state.

Delete a Snapshot
-----------------

To delete a snapshot, use the ``rbd snap rm`` command and specify the pool
name, the image name, and the snap name:

.. prompt:: bash $

   rbd snap rm {pool-name}/{image-name}@{snap-name}
	
For example:

.. prompt:: bash $

   rbd snap rm rbd/foo@snapname

.. note:: Ceph OSDs delete data asynchronously, so deleting a snapshot  does
   not immediately free up the capacity of the underlying OSDs. This process is
   known as "snaptrim", and is referred to as such in ``ceph status`` output.

Purge Snapshots
---------------

> To delete all snapshots, use the ``rbd snap purge`` command and the image name:

Reviewer: "and specify the pool name and the image name"

Create Snapshot
---------------

To create a snapshot, use the ``rbd snap create`` command and specify the pool

> name, the snap name, and the image name:

Reviewer: The suggested order was pool, image, snap -- it's the order in which they are specified on the command line, elsewhere on this page and also a natural progression from bigger to smaller.

---------------

> To create a snapshot with ``rbd``, specify the ``snap create`` option, the pool

Reviewer: `snap create` is not an option, it's a subcommand.  An example of an option is `--user` or `--no-progress`.  In another PR, we settled on not splitting the command (`rbd`) and the subcommand (`snap create`), so I'd phrase this as "To create a snapshot, use `rbd snap create` command and specify ...".

Author: Done.

name, and the image name:

.. prompt:: bash $

   rbd snap create {pool-name}/{image-name}@{snap-name}

For example:

.. prompt:: bash $

   rbd snap create rbd/foo@snapname
	

List Snapshots
--------------

> To list the snapshots of an image, specify the pool name and the image name:

Reviewer: Ditto -- "To list the snapshots of an image, use `rbd snap ls` command and specify ...".

Author: Done.

.. prompt:: bash $

   rbd snap ls {pool-name}/{image-name}

For example:

.. prompt:: bash $

   rbd snap ls rbd/foo

Roll back Snapshot
------------------

> To roll back to a snapshot with ``rbd``, specify the ``snap rollback`` option,

Reviewer: Ditto -- "To roll back an image to a snapshot, use `rbd snap rollback` command and specify ...".

Author: Done.

the pool name, the image name, and the snap name:

.. prompt:: bash $

   rbd snap rollback {pool-name}/{image-name}@{snap-name}

For example:

.. prompt:: bash $

   rbd snap rollback rbd/foo@snapname

.. note:: Rolling back an image to a snapshot means overwriting the current
   version of the image with data from a snapshot. The time it takes to execute
   a rollback increases with the size of the image. It is **faster to clone**
   from a snapshot **than to rollback** an image to a snapshot. Cloning from a
   snapshot is the preferred method of returning to a pre-existing state.

Delete a Snapshot
-----------------

> To delete a snapshot with ``rbd``, specify the ``snap rm`` subcommand, the pool

Reviewer: Ditto -- "To delete a snapshot, use `rbd snap rm` command and specify ...".

Author: Done.

name, the image name, and the snap name:

.. prompt:: bash $

   rbd snap rm {pool-name}/{image-name}@{snap-name}
	
For example:

.. prompt:: bash $

   rbd snap rm rbd/foo@snapname

.. note:: Ceph OSDs delete data asynchronously, so deleting a snapshot 

>    does not immediately free up the capacity of the underlying OSDs.

Reviewer: Perhaps also note that this asynchronous process is referred to as "snaptrim".  This term can be seen in `ceph status` output and other places.

Author: Done.

Purge Snapshots
---------------

> To delete all snapshots for an image with ``rbd``, specify the ``snap purge``

Reviewer: Ditto -- "To delete all snapshots, use `rbd snap purge` command and specify ...".

Author: Done.

           
               Parent                        Child

.. note:: The terms "parent" and "child" refer to a Ceph block device snapshot
   (parent) and the corresponding image cloned from the snapshot (child).
   These terms are important for the command line usage below.
   
Each cloned image (child) stores a reference to its parent image, which enables
the cloned image to open the parent snapshot and read it.

> A COW (copy-on-write) clone of a snapshot behaves exactly like any other Ceph

Reviewer: The abbreviation is already provided above and "copy-on-write" is used everywhere else on this page:
```suggestion
A copy-on-write clone of a snapshot behaves exactly like any other Ceph
```

Author: Done.

block device image. You can read to, write from, clone, and resize cloned
images. There are no special restrictions with cloned images. However, the

> copy-on-write clone of a snapshot depends on the snapshot, so you **MUST**

Reviewer: I don't think "must" needs to be capitalized here.  RBD won't let users to create a clone of a snapshot that isn't protected one way or another.

Reviewer: I'm not sure I'm following, what was the ask?

Author: @idryomov: I was going to ask you "Do we really need this bolded all-caps MUST in this text?"

But instead of asking, I thought I would wait to see if you complained about it. An opinion freely offered is worth much more than one solicited. I was pleased to see that you also thought that the MUST must go.

protect the snapshot before you clone it. The diagram below depicts this
process.

.. note:: Ceph supports the cloning of only "RBD format 2" images (that is,

>    images created with ``rbd create --image-format 2``). The kernel client

Reviewer: - `--image-format` has defaulted to `2`  for a very long time now.  Let's say "images created without specifying `--image-format 1` option" or something like that
- "The Linux kernel client ..."

Author: Done.

The cloned image has a reference to the parent snapshot, and includes the pool
ID, the image ID, and the snapshot ID. The inclusion of the pool ID means that
you may clone snapshots from one pool to images in another pool.

#. **Image Template:** A common use case for block device layering is to create

>    a master image and a snapshot that serves as a template for clones. For

Reviewer: Let's abandon the term master -- possible alternatives are base image, "golden" image, etc.

Author: I'm going with "base image".

   example: a user may create an image for a Linux distribution (for example,

>    Ubuntu 22.04) and create a snapshot for it. The user may occasionally update

Reviewer: Perhaps "a snapshot of it"?

Author: That is superior preposition choice.

Done.

>    Ubuntu 12.04) and create a snapshot for it. The user may occasionally update

Reviewer: I'd bump this version to something current while at it -- 22.04?

Author: Agreed and done.

   the image and create a new snapshot (by using such commands as ``sudo
   apt-get update``, ``sudo apt-get upgrade``, or ``sudo apt-get dist-upgrade``
   followed by ``rbd snap create``). As the image matures, the user can clone
   any one of the snapshots.

#. **Extended Template:** A more advanced use case includes extending a

>    template image that provides more information than a base image. For

Reviewer: Perhaps "to provide more information"?

Author: Done.

   example, a user may clone an image (for example, a VM template) and install
   other software (for example, a database, a content management system, an
   analytics system) and then snapshot the extended image, which may itself be
   updated just like the base image.

#. **Template Pool:** One way to use block device layering is to create a pool

>    that contains both (1) base images that act as templates and (2) snapshots

Reviewer: This needs to be reworded.  A snapshot is always attached to the image, a case where an image lives an one pool and its snapshot lives in another pool is not possible so "(1) ... and (2) ..." doesn't make sense from a technical perspective.  I'd simplify to "a pool that contains base images that act as templates and respective snapshots".

Author: > This needs to be reworded. A snapshot is always attached to the image, a case where an image lives an one pool and its snapshot lives in another pool is not possible so "(1) ... and (2) ..." doesn't make sense from a technical perspective. I'd simplify to "a pool that contains base images that act as templates and respective snapshots".

@idryomov I've written it thus:

**Template Pool:** One way to use block device layering is to create a pool
   that contains both base images that act as templates and their associated
   snapshots. You may then extend read-only privileges to users so that they
   may clone the snapshots without the ability to write or execute within the
   pool.

Does this properly express what it should?

Reviewer: I don't like "both" there.  One can't have base images without snapshots so "both" seems a bit redundant and misleading.  I'd either just drop "both" or replace "both" with "only" (or "just") to underscore that user-cloned images would need to go to a different pool as the point of the template pool is that it's read-only.

Author: @idryomov:

I've rewritten it thus:

> #. **Template Pool:** One way to use block device layering is to create a pool
>    that contains (1) base images that act as templates and (2) snapshots of
>    those templates. You may then extend read-only privileges to users so that
>    they may clone the snapshots even though they do not have permissions that
>    allow them to write or execute within the pool.

Let me know what you think.

Reviewer: LGTM

Author: Qapla.

>    that contains both (1) master images that act as templates and (2) snapshots

Reviewer: ditto -- master

Author: Done.

   of those templates. You may then extend read-only privileges to users so
   that they may clone the snapshots without the ability to write or execute
   within the pool.

#. **Image Migration/Recovery:** One way to use block device layering is to
   migrate or recover data from one pool into another pool.

Protecting a Snapshot
---------------------

Clones access the parent snapshots. All clones would break if a user

> inadvertently deleted the parent snapshot. To prevent data loss, you **MUST**

Reviewer: ditto -- unneeded "must" capitalization

Author: Done. (Removed.)

protect the snapshot before you can clone it:

.. prompt:: bash $

   rbd snap protect {pool-name}/{image-name}@{snapshot-name}

For example:

.. prompt:: bash $

>    rbd snap protect rbd/foo@-snapname

Reviewer: ```suggestion
   rbd snap protect rbd/foo@snapname
```

Author: Done.

>    rbd snap protect rbd/my-image@my-snapshot

Reviewer: Above, the example image and snapshot names were "foo" and "snapname" respectively.  Let's pick one and be consistent.

Author: Done. "foo@snapname" it shall be, throughout the document.

%s/my-image@my-snapshot/foo@snapname/g

Reviewer: If you chose to settle on "foo", I'd also change the cloned image name in cloning examples to "bar".

.. note:: You cannot delete a protected snapshot.

Cloning a Snapshot
------------------

> To clone a snapshot, specify the parent pool, the image, and the snapshot; and

Reviewer: Feel free to ignore, but "specify the parent pool, image and snapshot" read better than "specify the parent pool, the image, and the snapshot".  At least to me, a single "the" article made it clear that "parent" applies to all of pool, image and snapshot -- which is how it actually is.  With your wording, "parent" needs to be repeated twice more to make that clear.

Author: @idryomov: One of the areas that needs the most work in this text is the definite articles (and please don't get me wrong or take this as mordant criticism: the proper deployment of English articles is one of the slipperiest parts of English grammar). Is the following clause correct?

"To clone a snapshot, specify the parent pool, the parent image, and the parent snapshot;"

Reviewer: > @idryomov: One of the areas that needs the most work in this text is the definite articles (and please don't get me wrong or take this as mordant criticism: the proper deployment of English articles is one of the slipperiest parts of English grammar).

It wasn't me who wrote this, I would have done a worse job ;)

> Is the following clause correct?
> 
> "To clone a snapshot, specify the parent pool, the parent image, and the parent snapshot;"

Yes, it's correct.

Author: > It wasn't me who wrote this, I would have done a worse job ;)

You sell yourself short. 
 
> > Is the following clause correct?
> > "To clone a snapshot, specify the parent pool, the parent image, and the parent snapshot;"
> 
> Yes, it's correct.

This has been added to the doc.

snapshot; and also the child pool together with the image name. You must
protect the snapshot before you can clone it:

.. prompt:: bash $

>    rbd clone {pool-name}/{parent-image}@{snap-name} {pool-name}/{child-image-name}

Reviewer: ```suggestion
   rbd clone {pool-name}/{parent-image-name}@{snap-name} {pool-name}/{child-image-name}
```

Author: Done.

	
For example:

.. prompt:: bash $

>    rbd clone rbd/foo@snapname rbd/new-image

Reviewer: Given the use of "rbd/foo" for the parent, I'd use "rbd/bar" for the child:
```suggestion
   rbd clone rbd/foo@snapname rbd/bar
```

	

.. note:: You may clone a snapshot from one pool to an image in another pool.
   For example, you may maintain read-only images and snapshots as templates in
   one pool, and writeable clones in another pool.

Unprotecting a Snapshot
-----------------------

Before you can delete a snapshot, you must first unprotect it. Additionally,
you may *NOT* delete snapshots that have references from clones. You must

> flatten each clone of a snapshot before you can delete the snapshot:

Reviewer: "must flatten or delete each clone of a snapshot before you can unprotect the snapshot"

Author: Done.

snapshot:

.. prompt:: bash $

   rbd snap unprotect {pool-name}/{image-name}@{snapshot-name}

For example:

.. prompt:: bash $

   rbd snap unprotect rbd/foo@snapname

Listing Children of a Snapshot
------------------------------

> To list the children of a snapshot, execute the following:

Reviewer: "To list the children of a snapshot, use the `rbd children` command and specify ..." for consistency.

Author: Agreed and changed.

.. prompt:: bash $

   rbd children {pool-name}/{image-name}@{snapshot-name}

For example:

.. prompt:: bash $

   rbd children rbd/my-image@my-snapshot

Flattening a Cloned Image
-------------------------

Cloned images retain a reference to the parent snapshot. When you remove the

> reference to the parent snapshot from the child clone, you effectively

Reviewer: "child clone" is saying the same thing twice -- I'd just say "clone" here.

Author: Agreed. Done.

> "flatten" the image by copying the information from the snapshot to the clone.

Reviewer: "effectively "flatten" the clone by copying the data stored in the snapshot to the clone"

Author: Done.

The time it takes to flatten a clone increases with the size of the snapshot.

> To delete a snapshot, you must first flatten the child images:

Reviewer: "you must first flatten the child images (or delete them)"

Author: Done.

snapshot, you must first flatten the child images (or delete them):

.. prompt:: bash $

   rbd flatten {pool-name}/{image-name}

For example:

.. prompt:: bash $

>    rbd flatten rbd/new-image

Reviewer: ditto
```suggestion
   rbd flatten rbd/bar
```

Author: Done.

> .. note:: Since a flattened image contains all the information from the

Reviewer: "contains all data stored in the snapshot"

Author: Done.

   those templates. You may then extend read-only privileges to users so that
   they may clone the snapshots even though they do not have permissions that
   allow them to write or execute within the pool.

#. **Image Migration/Recovery:** One way to use block device layering is to
   migrate or recover data from one pool into another pool.

Protecting a Snapshot
---------------------

Clones access the parent snapshots. All clones would break if a user
inadvertently deleted the parent snapshot. To prevent data loss, you must 
protect the snapshot before you can clone it:

.. prompt:: bash $

   rbd snap protect {pool-name}/{image-name}@{snapshot-name}

For example:

.. prompt:: bash $

   rbd snap protect rbd/foo@snapname

.. note:: You cannot delete a protected snapshot.

Cloning a Snapshot
------------------

To clone a snapshot, specify the parent pool, the parent image, and the parent
snapshot; and also the child pool together with the image name. You must
protect the snapshot before you can clone it:

.. prompt:: bash $

   rbd clone {pool-name}/{parent-image-name}@{snap-name} {pool-name}/{child-image-name}
	
For example:

.. prompt:: bash $

>    rbd clone rbd/foo@snapname rbd/foo

Reviewer: ```suggestion
   rbd clone rbd/foo@snapname rbd/bar
```

	

.. note:: You may clone a snapshot from one pool to an image in another pool.
   For example, you may maintain read-only images and snapshots as templates in
   one pool, and writeable clones in another pool.

Unprotecting a Snapshot
-----------------------

Before you can delete a snapshot, you must first unprotect it. Additionally,
you may *NOT* delete snapshots that have references from clones. You must
flatten or delete each clone of a snapshot before you can unprotect the
snapshot:

.. prompt:: bash $

   rbd snap unprotect {pool-name}/{image-name}@{snapshot-name}

For example:

.. prompt:: bash $

   rbd snap unprotect rbd/foo@snapname

Listing Children of a Snapshot
------------------------------

To list the children of a snapshot, use the ``rbd children`` command and

> specify the following:

Reviewer: ```suggestion
specify the pool name, the image name, and the snap name:
```
